metals/src/main/scala/scala/meta/internal/metals/Compilations.scala

```diff
@@ -39,58 +46,97 @@ final class Compilations(
+      target: b.BuildTargetIdentifier
   ): Future[b.CompileResult] = {
-    compileBatch(targets)
+    compileBatch(target).map { results =>
+      results.getOrElse(target, new b.CompileResult(b.StatusCode.CANCELLED))
+    }
+  }
+
+  def compileFile(path: AbsolutePath): Future[b.CompileResult] = {
+    def empty = new b.CompileResult(b.StatusCode.CANCELLED)
+    for {
+      result <- {
+        expand(path) match {
+          case None => Future.successful(empty)
+          case Some(target) =>
+            compileBatch(target)
+              .map(res => res.getOrElse(target, empty))
+        }
+      }
+      _ <- compileWorksheets(Seq(path))
+    } yield result
   }
 
-  def compileFiles(paths: Seq[AbsolutePath]): Future[b.CompileResult] = {
+  def compileFiles(paths: Seq[AbsolutePath]): Future[Unit] = {
     val targets = expand(paths)
     for {
       result <- compileBatch(targets)
       _ <- compileWorksheets(paths)
-    } yield result
+    } yield ()
   }
 
-  def cascadeCompileFiles(paths: Seq[AbsolutePath]): Future[b.CompileResult] = {
+  def cascadeCompileFiles(paths: Seq[AbsolutePath]): Future[Unit] = {
     val targets =
       expand(paths).flatMap(buildTargets.inverseDependencies).distinct
     for {
-      result <- cascadeBatch(targets)
+      _ <- cascadeBatch(targets)
       _ <- compileWorksheets(paths)
-    } yield result
+    } yield ()
   }
 
   def cancel(): Unit = {
     compileBatch.cancelCurrentRequest()
     cascadeBatch.cancelCurrentRequest()
   }
 
-  def expand(paths: Seq[AbsolutePath]): Seq[b.BuildTargetIdentifier] = {
-    def isCompilable(path: AbsolutePath): Boolean =
+  def expand(path: AbsolutePath): Option[b.BuildTargetIdentifier] = {
+    val isCompilable =
       path.isScalaOrJava && !path.isDependencySource(workspace())
 
-    val compilablePaths = paths.filter(isCompilable)
-    val targets = compilablePaths.flatMap(buildTargets.inverseSources).distinct
+    if (isCompilable) {
+      val targetOpt = buildTargets.inverseSources(path)
 
-    if (targets.isEmpty && compilablePaths.nonEmpty) {
-      scribe.warn(s"no build target for: ${compilablePaths.mkString("\n  ")}")
-    }
+      if (targetOpt.isEmpty) {
+        scribe.warn(s"no build target for: $path")
+      }
 
-    targets
+      targetOpt
+    } else
+      None
   }
 
+  def expand(paths: Seq[AbsolutePath]): Seq[b.BuildTargetIdentifier] =
+    paths.flatMap(expand(_)).distinct
+
   private def compile(
       targets: Seq[b.BuildTargetIdentifier]
-  ): CancelableFuture[b.CompileResult] = {
-    val result = for {
-      connection <- buildServer()
-      if targets.nonEmpty
-    } yield compile(connection, targets)
-
-    result.getOrElse {
-      val result = new b.CompileResult(b.StatusCode.CANCELLED)
-      Future.successful(result).asCancelable
+  ): CancelableFuture[Map[BuildTargetIdentifier, b.CompileResult]] = {
+
+    val targetsByBuildServer = targets
+      .flatMap(t => buildServer(t).map(_ -> t).toSeq)
+      .groupBy(_._1)
```

Comment:
Maybe use full value names. It will be easier to figure out.  Plus, let's try to avoid `_1`, `_2` etc.

Reply:
Fixed.